proposals/dgl-operator-proposal.md

```diff
@@ -0,0 +1,248 @@
+      command:
+        - dglrun
+      args:
+        - '--graph-name'
+        - graphsage
+        - '--partition-entry-point'
+        - code/load_and_partition_graph.py
+        - '--num-partitions'
+        - '2'
+        - '--balance-train'
+        - '--balance-edges'
+        - '--train-entry-point'
+        - code/train_dist.py
+        - '--num-epochs'
+        - '1'
+        - '--batch-size'
+        - '1000'
+        - '--num-trainers'
+        - '1'
+        - '--num-samplers'
+        - '4'
+        - '--num-servers'
+        - '1'
+      volumeMounts:
+        - name: kube-volume
+          mountPath: /opt/kube
+        - name: config-volume
+          mountPath: /etc/dgl
+      imagePullPolicy: Always
+  restartPolicy: Never
+```
+
+### Resulting Workers
+
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: dgl-graphsage-worker-0
+spec:
+  volumes:
+    - name: shm-volume
+      emptyDir:
+        medium: Memory
+		# sizeLimit will auto set as the half of limited memory amount
+        sizeLimit: 10G
+    - name: config-volume
+      configMap:
+        name: dgl-graphsage-config
+        items:
+          - key: kubexec.sh
+            path: kubexec.sh
+            mode: 365
+          - key: ipconfig
+            path: ipconfig
+            mode: 292
+        defaultMode: 420
+    - name: kube-volume
+      emptyDir: {}
+  containers:
+    - name: dgl-graphsage
+      image: 'dgloperator/graphsage:v0.1.0'
+      command:
+        - sleep
+      args:
+        - 365d
+      ports:
+        - name: dglserver
+          containerPort: 30050
+          protocol: TCP
+      volumeMounts:
+        - name: shm-volume
+          mountPath: /dev/shm
+        - name: config-volume
+          mountPath: /etc/dgl
+        - name: kube-volume
+          mountPath: /opt/kube
+      imagePullPolicy: Always
+  restartPolicy: Never
+```
+
+## Design
+
+### Operator-wise
+
+1. Create a `ConfigMap` that contains:
+    1. `kubexec.sh`: A helper shell script that can invoke `kubectl exec` for remote execution
+    2. `ipconfig`: A text file lists the worker pods' name, ip, port (used in training phase of dgl server) and the available slots (GPUs) in each pod
+2. Create the RBAC resources (`Role`, `ServiceAccount`, `RoleBinding`) to allow remote execution (`pods/exec`).
+3. Create worker `Pods` that reaches the desired replicas. The command is set to sleep forever.
+4. Create launcher `Pods`. And `initContainer` (`kubectl-delivery`) in `launcher` pod takes the responsibility to:
+    1. wait for all worker pods are ready
+    2. `kubectl` binary is delivered to an `emptyDir` volume
+5. Take off main container of the launcher `Pod`, and run the `dglrun` command and arguments.
+    1. `dglrun`: A must-have binary file that can take the control of DGL distributed training.
+    2. The launcher is very light weight.
+6. When `DGLJob` finishes, it will clean all workers.
+
+![DGL Operator-wise Diagram](diagrams/dgl-operator-1.png)
```

Comment:
This is very similar to MPI operator's design. Any thoughts on how to refactoring and reuse what's in MPI operator?

Reply:
Leave my thoughts at your last comment (https://github.com/kubeflow/community/pull/512#discussion_r619407527)